File: web/src/types/investigation.ts
/**
 * Investigation Types
 *
 * TypeScript types mirroring the API schemas for investigations.
 */

export type InvestigationStatus =
  | 'pending'
  | 'planning'
  | 'executing'
  | 'correlating'
  | 'enriching'
  | 'completed'
  | 'failed'
  | 'cancelled';

export type ThreatLevel = 'critical' | 'high' | 'medium' | 'low' | 'info' | 'clean' | 'unknown';

export type EntityType =
  | 'domain'
  | 'ip_address'
  | 'email'
  | 'hash'
  | 'url'
  | 'person'
  | 'organization'
  | 'certificate'
  | 'vulnerability'
  | 'threat_actor'
  | 'malware'
  | 'campaign';

export interface TargetEntity {
  type: string;
  value: string;
  confidence: number;
}

export interface Finding {
  id: string;
  type: string;
  source: string;
  data: Record<string, unknown>;
  confidence: number;
  threat_level: ThreatLevel;
  timestamp: string;
}

export interface FindingDetail extends Finding {
  is_ioc: boolean;
  raw_response?: Record<string, unknown>;
}

export interface DiscoveredEntity {
  id: string;
  type: string;
  value: string;
  source_plugin: string;
  confidence: number;
  properties: Record<string, unknown>;
  discovered_at: string;
}

export interface ThreatAssessment {
  threat_level: ThreatLevel;
  confidence_score: number;
  is_malicious: boolean;
  threat_types: string[];
  indicators_of_compromise: number;
  attributed_actors: string[];
  mitre_techniques: string[];
  summary: string;
}

export interface InvestigationEvent {
  id: string;
  type: string;
  timestamp: string;
  data: Record<string, unknown>;
}

export interface Investigation {
  id: string;
  query: string;
  target: TargetEntity;
  status: InvestigationStatus;
  progress: number;
  current_stage: string;
  error?: string;

  created_at: string;
  started_at?: string;
  completed_at?: string;
  duration_seconds?: number;

  plugins_total: number;
  plugins_completed: number;
  plugins_failed: number;

  findings: Finding[];
  entities: DiscoveredEntity[];
  threat_assessment?: ThreatAssessment;
  events: InvestigationEvent[];
}

export interface InvestigationSummary {
  id: string;
  query: string;
  target: TargetEntity;
  status: InvestigationStatus;
  progress: number;
  current_stage: string;
  threat_level: ThreatLevel;
  findings_count: number;
  entities_count: number;
  created_at: string;
  completed_at?: string;
  duration_seconds?: number;
}

export interface InvestigationProgress {
  id: string;
  status: InvestigationStatus;
  progress: number;
  current_stage: string;
  plugins_completed: number;
  plugins_total: number;
}

export interface InvestigationCreate {
  query: string;
  entity_type?: EntityType;
  entity_value?: string;
  depth?: 'quick' | 'standard' | 'full';
}

export interface InvestigationListResponse {
  items: InvestigationSummary[];
  total: number;
  limit: number;
  offset: number;
}

export interface FindingListResponse {
  investigation_id: string;
  findings: FindingDetail[];
  total: number;
  by_source: Record<string, number>;
  by_threat_level: Record<string, number>;
}

// WebSocket event types
export type WSEventType =
  | 'connection.established'
  | 'investigation.started'
  | 'investigation.completed'
  | 'investigation.failed'
  | 'investigation.cancelled'
  | 'stage.changed'
  | 'plan.created'
  | 'plugin.started'
  | 'plugin.completed'
  | 'plugin.failed'
  | 'finding.discovered'
  | 'entity.discovered'
  | 'threat.detected'
  | 'progress.updated';

export interface WSMessage {
  type: WSEventType;
  investigation_id: string;
  timestamp: string;
  data: Record<string, unknown>;
}

File: spectre/api/schemas/investigation.py
"""
Investigation API Schemas

Pydantic models for investigation API requests and responses.
"""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from spectre.plugins.base import EntityType
from spectre.services.models import InvestigationStatus


class InvestigationCreate(BaseModel):
    """Request to start a new investigation."""

    query: str = Field(..., description="Target to investigate (domain, IP, email, hash, or natural language query)")
    entity_type: EntityType | None = Field(default=None, description="Explicit entity type (auto-detected if not provided)")
    entity_value: str | None = Field(default=None, description="Explicit entity value")
    depth: str = Field(default="standard", description="Investigation depth: quick, standard, or full")


class TargetEntityResponse(BaseModel):
    """Target entity in response."""

    type: str
    value: str
    confidence: float = 1.0


class FindingResponse(BaseModel):
    """A finding from a plugin."""

    id: str
    type: str
    source: str
    data: dict[str, Any]
    confidence: float
    threat_level: str
    timestamp: str


class DiscoveredEntityResponse(BaseModel):
    """A discovered entity."""

    id: str
    type: str
    value: str
    source_plugin: str
    confidence: float
    properties: dict[str, Any]
    discovered_at: str


class ThreatAssessmentResponse(BaseModel):
    """Threat assessment summary."""

    threat_level: str
    confidence_score: float
    is_malicious: bool
    threat_types: list[str]
    indicators_of_compromise: int
    attributed_actors: list[str]
    mitre_techniques: list[str]
    summary: str


class InvestigationEventResponse(BaseModel):
    """An investigation event."""

    id: str
    type: str
    timestamp: str
    data: dict[str, Any]


class InvestigationResponse(BaseModel):
    """Full investigation details."""

    id: str
    query: str
    target: TargetEntityResponse
    status: str
    progress: float
    current_stage: str
    error: str | None = None

    created_at: str
    started_at: str | None = None
    completed_at: str | None = None
    duration_seconds: float | None = None

    plugins_total: int
    plugins_completed: int
    plugins_failed: int

    findings: list[FindingResponse]
    entities: list[DiscoveredEntityResponse]
    threat_assessment: ThreatAssessmentResponse | None = None

    events: list[InvestigationEventResponse]


class InvestigationSummary(BaseModel):
    """Summary for list views."""

    id: str
    query: str
    target: TargetEntityResponse
    status: str
    progress: float
    current_stage: str
    threat_level: str
    findings_count: int
    entities_count: int
    created_at: str
    completed_at: str | None = None
    duration_seconds: float | None = None


class InvestigationListResponse(BaseModel):
    """Paginated list of investigations."""

    items: list[InvestigationSummary]
    total: int
    limit: int
    offset: int


class InvestigationProgress(BaseModel):
    """Lightweight progress update."""

    id: str
    status: str
    progress: float
    current_stage: str
    plugins_completed: int
    plugins_total: int


class WebSocketMessage(BaseModel):
    """WebSocket message format."""

    type: str
    investigation_id: str
    timestamp: str
    data: dict[str, Any]
